tools/serverpod_cli/lib/src/serverpod_packages_version_check/serverpod_packages_version_check.dart

```diff
@@ -0,0 +1,123 @@
+  static String approximateVersion(Version cliVersion) =>
+      'Package is defined with a version range (for example "^$cliVersion"). '
+      'Ensure full compatibility between the serverpod cli and serverpod '
+      'packages by defining a concrete package version. Prefer "$cliVersion" '
+      'over "^$cliVersion".';
+}
+
+List<SourceSpanException> performServerpodPackagesAndCliVersionCheck(
+  Version cliVersion,
+  Directory directory,
+) {
+  List<SourceSpanException> accumulatedWarnings = [];
+
+  var pubspecFiles = findPubspecsFiles(directory);
+  if (pubspecFiles.isEmpty) {
+    return accumulatedWarnings;
+  }
+
+  for (var pubspecFile in pubspecFiles) {
+    try {
+      var file = pubspecFile.readAsStringSync();
+      var pubspec = Pubspec.parse(file);
+      var yaml = loadYaml(file, sourceUrl: pubspecFile.uri) as YamlMap;
+      var warnings =
+          _validateServerpodPackagesVersion(cliVersion, pubspec, yaml);
+
+      accumulatedWarnings.addAll(warnings);
+    } catch (e) {
+      throw Exception(
+          'Error while parsing pubspec file: ${pubspecFile.path}: $e');
+    }
+  }
+
+  return accumulatedWarnings;
+}
+
+List<SourceSpanException> _validateServerpodPackagesVersion(
+  Version version,
+  Pubspec pubspec,
+  YamlMap yamlPubspec,
+) {
+  List<SourceSpanException> warnings = [];
+
+  var serverpodDependencies =
+      _getHostedServerpodDependencies(pubspec.dependencies);
+  if (serverpodDependencies.isNotEmpty) {
+    warnings.addAll(_validatePackageCompatibilities(
+        serverpodDependencies, version, yamlPubspec['dependencies']));
+  }
+
+  var serverpodDevDependencies =
+      _getHostedServerpodDependencies(pubspec.devDependencies);
+  if (serverpodDevDependencies.isNotEmpty) {
+    warnings.addAll(_validatePackageCompatibilities(
+        serverpodDependencies, version, yamlPubspec['devDependencies']));
+  }
+
+  return warnings;
+}
+
+List<MapEntry<String, HostedDependency>> _getHostedServerpodDependencies(
+  Map<String, Dependency> dependencies,
+) {
+  return dependencies.entries.fold([], (hostedDependencies, dependency) {
+    if (!dependency.key.startsWith('serverpod') ||
+        dependency.value is! HostedDependency) return hostedDependencies;
+
+    // Cast dependencies to HostedDependency type
+    return [
+      ...hostedDependencies,
+      MapEntry(dependency.key, dependency.value as HostedDependency)
+    ];
+  });
+}
+
+List<SourceSpanException> _validatePackageCompatibilities(
+  List<MapEntry<String, HostedDependency>> serverpodPackages,
+  Version cliVersion,
+  YamlMap packagesYaml,
+) {
+  List<SourceSpanException> packageWarnings = [];
+
+  for (var element in serverpodPackages) {
+    var packageName = element.key;
+    var package = element.value;
+    var packageYamlNode = packagesYaml.nodes[packageName];
+
+    if (packageYamlNode == null) {
+      throw SourceSpanException(
+          'Could not find package "$packageName" when inspecting as Yaml file',
```

Comment:
`Could not find package "$packageName" in pubspec file.`

As written now, it may not be a clear message to a user. Also, make sure all sentences end with a period in CLI output.